subsys/logging/log_backend_rtt.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2018 omSquare s.r.o.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log_backend.h>
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include <rtt/SEGGER_RTT.h>
+
+
+#if CONFIG_LOG_BACKEND_RTT_MODE_DROP
+
+#define DROP_MESSAGE "\nmessages dropped: 99 \r"
+#define DROP_MESSAGE_LEN (sizeof(DROP_MESSAGE) - 1)
+static const char *drop_msg = DROP_MESSAGE;
+static int drop_cnt = 0;
+static int drop_warn = 0;
+
+#else
+
+#define DROP_MESSAGE_LEN 0
+
+#endif /* CONFIG_LOG_BACKEND_RTT_MODE_DROP */
+
+#if CONFIG_LOG_BACKEND_RTT_BUFFER > 0
+
+#define RTT_LOCK()
+#define RTT_UNLOCK()
+static u8_t rtt_buf[CONFIG_LOG_BACKEND_RTT_BUFFER_SIZE];
+
+#else
+
+#define RTT_LOCK() SEGGER_RTT_LOCK()
+#define RTT_UNLOCK() SEGGER_RTT_UNLOCK()
+
+#endif /* CONFIG_LOG_BACKEND_RTT_BUFFER > 0 */
+
+static u8_t line_buf[CONFIG_LOG_BACKEND_RTT_MESSAGE_SIZE + DROP_MESSAGE_LEN];
+static u8_t *line_pos;
+static u8_t char_buf;
+static int panic_mode;
+
+static int msg_out(u8_t *data, size_t length, void *ctx);
+
+static int line_out(u8_t data);
+
+static void log_backend_rtt_flush(void);
+
+static int log_backend_rtt_write(void);
+
+static int log_backend_rtt_panic(u8_t *data, size_t length);
+
+static int msg_out(u8_t *data, size_t length, void *ctx)
+{
+	(void) ctx;
+	u8_t *pos;
+
+	if (panic_mode) {
+		return log_backend_rtt_panic(data, length);
+	}
+
+	for (pos = data; pos < data + length; pos++) {
+		if (line_out(*pos)) {
+			break;
+		}
+	}
+
+	return (int) (pos - data);
+}
+
+static int line_out(u8_t data)
+{
+	if (data == '\r') {
+		if (log_backend_rtt_write()) {
+			return 1;
+		}
+		line_pos = line_buf;
+		return 0;
+	}
+
+	if (line_pos < line_buf + sizeof(line_buf) - 1) {
+		*line_pos++ = data;
+	}
+
+	/* not enough space in line buffer, we have to wait for EOL */
+	return 0;
+}
+
+#if CONFIG_LOG_BACKEND_RTT_MODE_DROP
```

Comment:
if possible try to avoiding #if you can do something like:
```
int log_backend_rtt_write(void)
{
    if (IS_ENABLED(CONFIG_LOG_BACKEND_RTT_MODE_DROP)) { drop_write()}
   else block_write()
}
```

Then compiler will cut what is not used but will compile everything. As block approach is the default one, i'm affraid that at some point drop will not compile because it won't be that frequently used (e.g. segger api changes).

Reply:
NP, sounds reasonable, will do, however in this case intersection of both block uses same set of SEGGER API.
Do you still want for me to change that?

Reply:
Segger API change was only the example. It's rather about the approach and not that particular case. The thing is that if you use IS_ENABLED macro you get all your code compiled even for not used options.

Reply:
Agree, better safe than sorry. Will do.

Reply:
Fixed by https://github.com/zephyrproject-rtos/zephyr/pull/10659/commits/06b91b792826215048f7166a032be0fab4d71a43